Upgrade/Install: Improve notice for missing dependencies.

costdev opened the pull request:
This PR changes the messaging for the notice displayed when there are missing dependencies.

In addition to removing the word "dependencies", which may not be understood by the average user, a link to the plugins page with some direction has been added, and the notice is now a `.notice-warning` rather than the previous `.notice-info`.

Furthermore, an error notice is replaces the previous "Note:" displayed in a dependent's row to inform the user that the plugin either cannot be activated, or is active but will not function correctly.

### Plugins > Installed plugins

#### Top: Before
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/47ee0554-eca8-4d84-af26-701230680805)

#### Top: After
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/ced30e5c-04d4-4410-803d-e2003f8ca187)

#### Plugin Row: Before
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/1914acc4-8169-4f99-8a5b-3c7e44dbb100)

#### Plugin Row: After

##### Inactive dependent
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/91325102-e796-4bda-ae19-9c26778ef4cf)

##### Active dependent and dependencies deleted from `wp-content/plugins`
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/3bbebad4-c75f-45c1-934b-e3239ad06cf7)

### Plugins > Add New

#### Before
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/970fd253-f9b0-4b77-add3-7301c82c53fd)

#### After
![image](https://github.com/WordPress/wordpress-develop/assets/79332690/a201bea9-e591-474b-a148-cc469edac50e)

Trac ticket: https://core.trac.wordpress.org/ticket/60465

costdev committed:
Upgrade/Install: Improve notice for missing dependencies.

joedolson commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -359,10 +359,16 @@ public static function get_dependency_data( $slug ) {
-				__( 'There are additional plugin dependencies that must be installed.' ),
+				sprintf(
+					/* translators: 1: The start of a link to the plugins page, 2: The end of the link. */
+					__( 'Some required plugins are missing. Please %1$sreview your website&#8217;s plugins%2$s for plugins that cannot be activated yet.' ),
```
Is the word "website's" really needed here? I think that "Some required plugins are missing. Please review your plugins for plugins that cannot be activated." would be sufficient. I think that 'website' can be assumed, since there is no other possible context, and the time frame of "the present" can also be assumed.

costdev replied:
Thanks Joe! Done in [fd17150](https://github.com/WordPress/wordpress-develop/pull/6070/commits/fd1715074234b1ff0727faee20c335fa0d09b0af).

costdev committed:
Remove "website's" and "yet" per review feedback.

afragen commented:
How about "Some required plugins are missing. Please review your plugins for those that cannot be activated." With link as appropriate.

costdev commented:
@afragen Ah I see, to reduce the repetition of the word "plugins". @joedolson what do you think?

desrosj commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -359,10 +359,16 @@ public static function get_dependency_data( $slug ) {
-				__( 'There are additional plugin dependencies that must be installed.' ),
+				sprintf(
+					/* translators: 1: The start of a link to the plugins page, 2: The end of the link. */
+					__( 'Some required plugins are missing. Please %1$sreview your plugins%2$s for plugins that cannot be activated.' ),
```
I agree with the feedback that plugins appears too many times.

I'd love if we had a more concrete step for the user. "Review your plugins" is not the greatest direction. What are they looking for?

Active and inactive plugins have different background colors, and when a site is in recovery mode, there is a third styling to call attention to the problematic rows.

![Screenshot 2024-02-12 at 10 12 41 AM](https://github.com/WordPress/wordpress-develop/assets/359867/01a85051-4431-442c-9c05-931c82b9ab13)

What about styling plugins with unmet dependencies differently?

Another option could be making the greyed out `Active` link was changed to say `Missing requirements`?

What would be really nice is a link to a list of all plugins missing dependencies. I opened a suggestion for that in [Trac-60507](https://core.trac.wordpress.org/ticket/60507). Then it could read "View plugins with missing requirements" and link to that filtered view.

afragen replied:
I am far from an accessibility expert but I thought color change wasn't supposed to be a great differentiator.

desrosj commented:
@joedolson Another a11y question. Should this notice have a "Warning:" prefix? Or is the notice on it's own without a prefix OK?

costdev committed:
Remove third instance of "plugins".

costdev committed:
Replace Dependencies "Note" with an error notice.

afragen reviewed (approved):
If all the accessibility stuff is good, this looks good.

swissspidy commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -369,10 +369,16 @@ public static function get_dependency_data( $slug ) {
 			wp_admin_notice(
-				__( 'There are additional plugin dependencies that must be installed.' ),
+				sprintf(
+					/* translators: 1: The start of a link to the plugins page, 2: The end of the link. */
+					__( 'Some required plugins are missing. Please %1$sreview your plugins%2$s for those that cannot be activated.' ),
+					'<a href="' . esc_url( $plugins_url ) . '">',
+					'</a>'
```
We don't really do that when there are links. Just use HTML.

```suggestion
					/* translators: %s: Link to the plugins page. */
					__( 'Some required plugins are missing. Please <a href="%s">review your plugins</a> for those that cannot be activated.' ),
					esc_url( $plugins_url ),
```

costdev committed:
Add feedback from review

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

swissspidy commented:
@costdev note the merge conflict now

costdev committed:
Merge branch 'trunk' into plugin_dependencies_improve_notice_messaging

costdev committed:
Add `.php` to the plugins admin URL.

costdev committed:
Account for active/inactive dependents in top notice.

Don't mention "for plugins that cannot be activated."

There is a notice on each dependent plugin row that has a problem.

costdev committed:
Tailor plugin row notice to the dependent's current state.

costdev committed:
Remove errant comma.

costdev committed:
Add `or inactive` to the top notice.

costdev commented:
@desrosj @afragen @swissspidy I've updated the top notice so that:
1. It says "required plugins that are missing or inactive" for consistency.
2. Doesn't say "for plugins that cannot be activated", as the issue may (and likely will) actually be plugins that were already active when their dependencies were deleted or deactivated through some other means.

When users click the link to `plugins.php`, they'll see a notice in a problematic dependent's row which changes based on the dependent's active/inactive status. See the PR's description for updated screenshots.

Thoughts? 🙂

andreilupu commented:
Hi all,

I want to bring up the point that the plugin card might get too crowded with this notice(that's just my opinion).

If we take into consideration an update notice (and who knows what else people bring extra in that card) it would look like this
![Screenshot 2024-03-03 at 15 37 04](https://github.com/WordPress/wordpress-develop/assets/1893980/9c1e2f0d-64d4-47bf-99f2-5d5633d08655)

Personally, I share @desrosj thoughts from [track#60465](https://core.trac.wordpress.org/ticket/60465) I'm not sure if that notice adds much value and it might not be needed. A contextual help tab proposed in [track#60466](https://core.trac.wordpress.org/ticket/60466) could add more info than that notice.

I would argue that a plugin required by another is self-explanatory, it is clear that you cannot deactivate the main plugin as long as the dependency is active.

Maybe a little bit of visual focus might help.
![Screenshot 2024-03-03 at 15 46 05](https://github.com/WordPress/wordpress-develop/assets/1893980/43e84e22-9c59-4855-b1d7-88b5595267a6)

Congrats for the work on this feature, I can't wait for it to land in production.

desrosj commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -1567,19 +1567,35 @@ protected function add_dependencies_to_dependent_plugin_row( $dependent ) {
+		$error_message = '';
+		if ( WP_Plugin_Dependencies::has_unmet_dependencies( $dependent ) ) {
+			if ( $is_active ) {
+				$error_message = __( 'This plugin is active but will not function correctly because required plugins are missing or inactive.' );
```
```suggestion
				$error_message = __( 'This plugin is active but may not function correctly because required plugins are missing or inactive.' );
```
It's possible that the plugin could still work correctly, but not guaranteed.

desrosj commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -369,10 +369,15 @@ public static function get_dependency_data( $slug ) {
-				__( 'There are additional plugin dependencies that must be installed.' ),
+				sprintf(
+					/* translators: %s: Link to the plugins page. */
+					__( 'Some required plugins are missing or inactive. Please <a href="%s">review your plugins</a>.' ),
```
Will this notice display anywhere other than at the top of the Plugins page? If not, I don't think a link is necessary. The only exception, I think, would be if you're on a site within a network. Then I think it makes sense to link to the network admin plugin page, but only if the user has that capability.

costdev replied:
It'll appear on `Plugins > Installed Plugins` and `Plugins > Add New`

desrosj commented:
Thanks for working on this, all! I added a few comments that I wouldn't necessarily consider blockers. This is a good refinement and a step in the right direction for 6.5.

desrosj commented:
> I want to bring up the point that the plugin card might get too crowded with this notice(that's just my opinion).

I do agree that if a plugin has several dependencies, it will quickly crowd the plugin card view. But, since it will take a bit of time for plugins to fully embrace this, I think that we can circle back to this in a future iteration.

costdev committed:
Update src/wp-admin/includes/class-wp-plugins-list-table.php

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

costdev commented:
@desrosj Are we good to go for commit with this one?

desrosj committed:
Only show link when necessary.

Also, add some checks for proper access for multisite setups.

desrosj commented:
I made similar changes here to the ones in response to my [comment on the other PR](https://github.com/WordPress/wordpress-develop/pull/6074#discussion_r1510973534).

> I don't know that it's useful or helpful to link to the network admin URL if the user does not have the ability to access that page.

Also, the Manage Plugins link will no longer display when looking at the actual plugins page. The user is already on that page, there's no need to link them.

I think this is good to go, but would love a sanity check from someone else first.

peterwilsoncc commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -369,10 +369,29 @@ public static function get_dependency_data( $slug ) {
+
+			if ( is_multisite() ) {
+				if ( current_user_can( 'manage_network_plugins' ) ) {
+					$error_message .= ' ' . sprintf(
+						/* translators: %s: Link to the network plugins page. */
+						esc_url( network_admin_url( 'plugins.php' ) )
+					);
```
This is missing the string in which the URL is placed.

It would be good to check with an i18n expert that the string can be concatenated like this and still be translated. If not we might need a few more if statements. @swissspidy Are you able to help advise here?

desrosj replied:
Good catch. I fixed the missing string.

Good call out on the concatenation. We did [confirm this is acceptable on the other PR](https://github.com/WordPress/wordpress-develop/pull/6074#discussion_r1511837905).

peterwilsoncc replied:
> Good call out on the concatenation. We did confirm....

Thanks. As a monolingual English speaker I like to check these things because it's easy for me to misunderstand.

peterwilsoncc reviewed (commented):
I noticed a missing string and have asked for some help with translatability.

desrosj committed:
Add missing string.

desrosj commented:
Thanks all! Committed in https://core.trac.wordpress.org/changeset/57769.
